src/api.ts

```diff
@@ -195,69 +201,245 @@ function createApi({
         throw new NotFoundError(id)
       }
-      await tilestore.putTileJSON(tilejson)
-      return { ...tilejson, id, tiles: [getTileUrl(id)] }
+
+      const tilesFieldValue = [getTileUrl(id)]
+
+      db.prepare<{
+        id: string
+        format: TileJSON['format']
+        tilejson: string
+        upstreamTileUrls: string
+      }>(
+        'UPDATE Tileset SET ' +
+          'tilejson = :tilejson, ' +
+          'format = :format, ' +
+          'upstreamTileUrls = :upstreamTileUrls ' +
+          'WHERE id = :id'
+      ).run({
+        id,
+        format: tilejson.format,
+        tilejson: JSON.stringify({ ...tilejson, tiles: tilesFieldValue }),
+        upstreamTileUrls: JSON.stringify(tilejson.tiles),
+      })
+
+      const result = {
+        ...tilejson,
+        tiles: tilesFieldValue,
+        id,
+      }
+
+      return result
     },
 
     async listTilesets() {
-      const tilesetIds = Array.from(context.tilestores.keys())
-      return Promise.all(tilesetIds.map((id) => api.getTileset(id)))
+      return db
+        .prepare('SELECT id, tilejson FROM Tileset')
+        .all()
+        .map(({ id, tilejson }: { id: string; tilejson: string }) => ({
+          ...JSON.parse(tilejson),
+          id,
+        }))
     },
 
     async getTileset(id) {
-      const tilestore = context.tilestores.get(id)
-      if (!tilestore) {
+      const row:
+        | { tilejson: TileJSON; etag?: string; upstreamUrl?: string }
+        | undefined = db
+        .prepare('SELECT tilejson, etag, upstreamUrl FROM Tileset WHERE id = ?')
+        .get(id)
+
+      if (!row) {
         throw new NotFoundError(id)
       }
-      return {
-        ...(await tilestore.getTileJSON()),
-        id,
+
+      async function fetchOnlineResource(url: string, etag?: string) {
+        const { data } = await upstreamRequestsManager.getUpstream<TileJSON>({
+          url,
+          etag,
+          responseType: 'json',
+        })
+
+        if (data) api.putTileset(id, data)
+      }
+
+      if (row.upstreamUrl) {
+        fetchOnlineResource(row.upstreamUrl, row.etag).catch(() => {})
       }
+
+      return { ...row.tilejson, id }
     },
 
     async getTile({ tilesetId, zoom, x, y }) {
-      const tilestore = context.tilestores.get(tilesetId)
-      if (!tilestore) {
-        throw new NotFoundError(tilesetId)
+      const tilejson = await api.getTileset(tilesetId)
+
+      const upstreamTileUrl = getUpstreamTileUrl(tilejson, { zoom, x, y })
+
+      const quadKey = tileToQuadkey([x, y, zoom])
+
+      let tile: { data: Buffer; etag?: string } | undefined = db
+        .prepare<{
+          tilesetId: string
+          quadKey: string
+        }>(
+          'SELECT TileData.data as data, Tileset.etag as etag FROM TileData ' +
+            'JOIN Tile ON TileData.tileHash = Tile.tileHash ' +
+            'JOIN Tileset ON Tile.tilesetId = Tileset.id ' +
+            'WHERE Tileset.id = :tilesetId AND Tile.quadKey = :quadKey'
```

Comment:
tbh i have no idea if this query works as expected. in theory it does, but having an e2e test to verify would be ideal